README.md
pastecapi
=========

Lightweight Node.js module for the Pastec image recognition API.
Both the opensource and Hosted API are supported.

[![npm](https://img.shields.io/npm/v/pastecapi.svg?maxAge=3600)](https://github.com/fvdm/nodejs-pastecapi/blob/master/CHANGELOG.md)
[![Build Status](https://travis-ci.org/fvdm/nodejs-pastecapi.svg?branch=master)](https://travis-ci.org/fvdm/nodejs-pastecapi)
[![Dependency Status](https://gemnasium.com/badges/github.com/fvdm/nodejs-pastecapi.svg)](https://gemnasium.com/github.com/fvdm/nodejs-pastecapi#runtime-dependencies)
[![Coverage Status](https://coveralls.io/repos/github/fvdm/nodejs-pastecapi/badge.svg?branch=master)](https://coveralls.io/github/fvdm/nodejs-pastecapi?branch=master)

* [Node.js](https://nodejs.org)
* [Pastec](http://pastec.io)
* [API documentation](http://pastec.io/doc#api)


Example
-------

```js
var pastec = require ('pastecapi') ({
  endpoint: 'https://my-pastec-server:4212'
});

// Add image signature to index
pastec.addImage ('./advertising.jpg', console.log);

// Match image against index
pastec.searchIndex ('./logo.jpg', console.log);
```


Installation
------------

`npm install pastecapi`


Configuration
-------------
**( [config] )**

**( [endpoint], [authkey], [timeout] )**


The setup function can take either a config object with settings or arguments.


setting  | type   | required | default                 | description
:--------|:-------|:---------|:------------------------|:-------------------
endpoint | string | no       | `http://localhost:4121` | API endpoint
indexid  | string | no       |                         | Hosted API Index ID
authkey  | string | no       |                         | API auth key, from `--auth-key` flag or Hosted API
timeout  | number | no       | `5000`                  | Wait timeout in ms


```js
// Example 1 - All defaults
var pastec = require ('pastecapi') ();

// Example 2 - Using Hosted API (Pastec.io)
var pastec = require ('pastecapi') ({
  endpoint: 'https://api.pasted.io',
  indexid: 'your index id'
  authkey: 'your authentication key'
});
```


Methods
-------

All methods take a callback function to process the results.

```js
function myCallback (err, data) {
  if (err) {
    console.log (err);
    return;
  }

  console.log (data.image_ids);
}

pastec.searchIndex ('~/image.jpg', myCallback);
```

#### Errors

message                      | description              
:----------------------------|:--------------------------------------------------------
request failed               | Request error occured, see `err.statusCode`, `err.error`
invalid authkey              | Your `authkey` or `indexid` is invalid
endpoint does not allow ping | The API does not accept [.ping](#ping) method


### .addImage
**( image, imageId, callback )**

Upload image to server.


param    | type             | required | description
:--------|:-----------------|:---------|:-----------------------------
image    | buffer or string | yes      | Buffer or local path to image
imageId  | number           | yes      | Index ID for image signature
callback | function         | yes      | Process result


```js
pastec.addImage ('~/image.jpg', 123, console.log);
```


### .deleteImage
**( imageId, callback )**

Remove image signature from index.


param    | type     | required | description
:--------|:---------|:---------|:----------------------------
imageId  | number   | yes      | Index ID for image signature
callback | function | yes      | Process result


```js
pastex.deleteImage (123, console.log);
```


### .searchIndex
**( image, console.log )**

Upload image to server and match its signature against the index.


param    | type             | required | description
:--------|:-----------------|:---------|:-----------------------------
image    | buffer or string | yes      | Buffer or local path to image
callback | function         | yes      | Process result


```js
pastec.searchIndex ('~/image.jpg', console.log);
```


### .loadIndex
**( filepath, callback )**

Load another index from file on the server.


param    | type     | required | description
:--------|:---------|:---------|:---------------------------
filepath | string   | yes      | Path to .dat file on server
callback | function | yes      | Process result


```js
pastec.loadIndex ('/var/lib/pastec/index.dat', console.log);
```


### .writeIndex
**( filepath, callback )**

Write current index to a file on the server.


param    | type     | required | description
:--------|:---------|:---------|:---------------------------
filepath | string   | yes      | Path to .dat file on server
callback | function | yes      | Process result


```js
pastec.writeIndex ('/var/lib/pastec/index.dat', console.log);
```


### .clearIndex
**( callback )**

Clear the current index on the server.


param    | type     | required | description
:--------|:---------|:---------|:--------------
callback | function | yes      | Process result


```js
pastec.clearIndex (console.log);
```


### .ping
**( callback )**

Ping the server to test API access.

**Note:** `.ping` is not available on the Hosted API at `api.pastec.io`.


param    | type     | required | description
:--------|:---------|:---------|:--------------
callback | function | yes      | Process result


```js
pastec.ping (function (err, data) {
  if (err) { return console.log (err); }

  if (data.type === 'PONG') {
    console.log ('Yay it works!');
  } else {
    console.log ('Oh no...');
    console.log (data);
  }
});
```


License
-------

This is free and unencumbered software released into the public domain.

Anyone is free to copy, modify, publish, use, compile, sell, or
distribute this software, either in source code form or as a compiled
binary, for any purpose, commercial or non-commercial, and by any
means.

In jurisdictions that recognize copyright laws, the author or authors
of this software dedicate any and all copyright interest in the
software to the public domain. We make this dedication for the benefit
of the public at large and to the detriment of our heirs and
successors. We intend this dedication to be an overt act of
relinquishment in perpetuity of all present and future rights to this
software under copyright law.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR
OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
OTHER DEALINGS IN THE SOFTWARE.

For more information, please refer to <http://unlicense.org>


Author
------

[Franklin van de Meent](https://frankl.in)


CHANGELOG.md
#### 1.3.1 (2016-6-14)

##### Documentation Changes

* **methods:** Mention ping unavailable at hosted API ([cbed8d8e](https://github.com/fvdm/nodejs-pastecapi/commit/cbed8d8ebd116314346b06f9461b20588967e7a8))
* **configuration:**
  * Mention hosted API for authkey ([171636cd](https://github.com/fvdm/nodejs-pastecapi/commit/171636cd92655689fd9097b2f43f7285f067bd0a))
  * Add indexid setting ([7e331dd1](https://github.com/fvdm/nodejs-pastecapi/commit/7e331dd1b7f2a55fb2e95b76ce4e2490f48d57ef))
  * Clarify configs are examples ([22c96774](https://github.com/fvdm/nodejs-pastecapi/commit/22c967745be3d78167cc2645fed1915cc99ad143))
  * Use object config ([224c6ebc](https://github.com/fvdm/nodejs-pastecapi/commit/224c6ebcd8960a78528c4caeb741ac3422160948))
* **readme:** Add error messages ([6f139bd3](https://github.com/fvdm/nodejs-pastecapi/commit/6f139bd3ead8261f601413ad20fd36400043a856))
* **intro:** Mention opensource and hosted APIs ([4a922837](https://github.com/fvdm/nodejs-pastecapi/commit/4a922837cb4bfc63bdcbcc57aa3b1f6e2680e1a6))

### 1.3.0 (2016-6-13)

##### Chores

* **package:** Update eslint dev dep ([d9a97743](https://github.com/fvdm/nodejs-pastecapi/commit/d9a977435c2848636a8aacd3596fb6693b6e6f55))

##### New Features

* **interface:** Add .config to interface ([2f1858c3](https://github.com/fvdm/nodejs-pastecapi/commit/2f1858c3fc35c06a755846e2885efc9a9070dcce))
* **setup:**
  * Add indexid configuration setting ([3d8d0803](https://github.com/fvdm/nodejs-pastecapi/commit/3d8d080378d743e8e9fba8be6f1131fd68bf0ff1))
  * Allow config object instead of arguments ([5da2b7bb](https://github.com/fvdm/nodejs-pastecapi/commit/5da2b7bb034c0637d2a2e9fcb6809e8117a5d245))

##### Bug Fixes

* **cleanup:** Remove trailing whitespace ([0bddeeed](https://github.com/fvdm/nodejs-pastecapi/commit/0bddeeedb4a365eb317239e61d9f49adad417d23))
* **http:**
  * Ping error only for hosted API ([5b1c1e94](https://github.com/fvdm/nodejs-pastecapi/commit/5b1c1e94c463f8d4ca4b39690816d50cf5bad88b))
  * Catch another authkey error ([7806602f](https://github.com/fvdm/nodejs-pastecapi/commit/7806602faab5d62c7ad301800589c52566dde125))
  * Check options.json existence ([d5562284](https://github.com/fvdm/nodejs-pastecapi/commit/d5562284f9783845e3c7dc46c87f0b1ba173e328))
  * Hosted API does not allow ping ([3b3c320b](https://github.com/fvdm/nodejs-pastecapi/commit/3b3c320ba49d3b214564a147ae0c8f2572c2231a))
  * Fixed path index and invalid authkey error ([edea3fce](https://github.com/fvdm/nodejs-pastecapi/commit/edea3fce11c27aac23754818e92100c0c373a562))
  * Fixed syntax typo) ([ef87698c](https://github.com/fvdm/nodejs-pastecapi/commit/ef87698ca00d377243c551017d57bf6520589ea2))

##### Refactors

* **style:** Restore cleaner returns ([04462f49](https://github.com/fvdm/nodejs-pastecapi/commit/04462f49c5d1357ca87619e40de6361436bfcdd5))

##### Tests

* **style:** Replace const/let with var ([61baf41e](https://github.com/fvdm/nodejs-pastecapi/commit/61baf41e4608286864527eb30162cd5da09bae16))
* **tests:**
  * Run both opensource and hosted APIs ([6fdf6f68](https://github.com/fvdm/nodejs-pastecapi/commit/6fdf6f68dd5e2d6253b287d68ddf0e102275de44))
  * Add Setup with arguments ([887641f5](https://github.com/fvdm/nodejs-pastecapi/commit/887641f5d54b387bb62f945fbd62c624b8283b3d))
* **fix:**
  * warn about invalid authkey response ([3fd33fc2](https://github.com/fvdm/nodejs-pastecapi/commit/3fd33fc228457afeb5280187ea236f384a500cdf))
  * Use writeIndex instead of ping for errors ([4749b050](https://github.com/fvdm/nodejs-pastecapi/commit/4749b050c917a4a6cda2107ef05788726972e67c))
  * Fixed syntax typo ([01d91882](https://github.com/fvdm/nodejs-pastecapi/commit/01d91882d85347ce2aae4340cca5f13179a39d38))
* **setup:** Add PASTEC_INDEXID, use object config ([f7e1f5f1](https://github.com/fvdm/nodejs-pastecapi/commit/f7e1f5f1986ea56a8f50aa32fc283bcbb9510328))

#### 1.2.3 (2016-6-10)

##### Documentation Changes

* **badges:** Fixed dependencies deeplink ([1d885a47](https://github.com/fvdm/nodejs-pastecapi/commit/1d885a47d6702b3cb6f154d5fc928ede3a730c06))

#### 1.2.2 (2016-6-10)

##### Documentation Changes

* **readme:** Clarify configuration syntax ([8d2de9fd](https://github.com/fvdm/nodejs-pastecapi/commit/8d2de9fd8fa06e1c829369e4f99bc4e9958435d4))

#### 1.2.1 (2016-6-10)

##### Documentation Changes

* **readme:** Add configuration section ([51b583fe](https://github.com/fvdm/nodejs-pastecapi/commit/51b583fecb4c176f8f70e5fa76d43c199d7941e6))

### 1.2.0 (2016-6-10)

##### Chores

* **jsdoc:** Describe authkey setting ([2bd3250b](https://github.com/fvdm/nodejs-pastecapi/commit/2bd3250b297a5d8952376b21a5c17e12eb824a83))
* **package:** Update dependency versions ([721c7548](https://github.com/fvdm/nodejs-pastecapi/commit/721c7548a6fc76c288a3707830eeb57ee2fba026))

##### Documentation Changes

* **badges:**
  * Add coverage status ([4ea5f0de](https://github.com/fvdm/nodejs-pastecapi/commit/4ea5f0de33b4915aad64028eea5348bdc12a3154))
  * Deeplink Gemnasium to dependencies tab ([1cf59c05](https://github.com/fvdm/nodejs-pastecapi/commit/1cf59c05efa1ee82cf28c84628eb643a0d57716b))
  * Add npm version for changelog ([dfa280a2](https://github.com/fvdm/nodejs-pastecapi/commit/dfa280a230934ea6a6fbc7405711bb4e8ebd0117))

##### New Features

* **http:** Add authkey support ([cdcde5d2](https://github.com/fvdm/nodejs-pastecapi/commit/cdcde5d273ad28fb71bf4e8befbff97d88f31bd0))

##### Bug Fixes

* **style:** Consistent returns ([219316f7](https://github.com/fvdm/nodejs-pastecapi/commit/219316f79cb5fd7580acfbd0a402a15f4088a42a))
* **main:** Cleaner, consistent returns ([e25e38e8](https://github.com/fvdm/nodejs-pastecapi/commit/e25e38e80d41d0748e3364ea125b51a415678a67))
* **errors:** Catch invalid authkey error ([bf506260](https://github.com/fvdm/nodejs-pastecapi/commit/bf506260c4e5c4f335e0e7efeec5086ccc2927aa))

##### Other Changes

* **undefined:**

test.js
/*
Name:           pastecapi - test.js
Author:         Franklin van de Meent (https://frankl.in)
Source & docs:  https://github.com/fvdm/nodejs-pastecapi
Feedback:       https://github.com/fvdm/nodejs-pastecapi/issues
License:        Unlicense (Public Domain, see LICENSE file)
                (https://github.com/fvdm/nodejs-pastecapi/raw/develop/LICENSE)
*/


const fs = require ('fs');
const dotest = require ('dotest');
const app = require ('./');
const path = require ('path');
const dir = path.dirname (module.filename);

const opensource = {
  endpoint: process.env.OPENSOURCE_ENDPOINT || 'http://localhost:4212',
  indexid: process.env.OPENSOURCE_INDEXID || null,
  authkey: process.env.OPENSOURCE_AUTHKEY || null,
  timeout: process.env.OPENSOURCE_TIMEOUT || 5000
};

const hosted = {
  endpoint: process.env.HOSTED_ENDPOINT || 'https://api.pastec.io',
  indexid: process.env.HOSTED_INDEXID || null,
  authkey: process.env.HOSTED_AUTHKEY || null,
  timeout: process.env.HOSTED_TIMEOUT || 5000
};

let pastec = app (opensource);


// Tests
dotest.add ('Module', function (test) {
  test ()
    .isFunction ('fail', 'exports', app)
    .isObject ('fail', 'module', pastec)
    .isObject ('fail', '.config', pastec && pastec.config)
    .isFunction ('fail', '.ping', pastec && pastec.ping)
    .isFunction ('fail', '.writeIndex', pastec && pastec.writeIndex)
    .isFunction ('fail', '.clearIndex', pastec && pastec.clearIndex)
    .isFunction ('fail', '.loadIndex', pastec && pastec.loadIndex)
    .isFunction ('fail', '.searchIndex', pastec && pastec.searchIndex)
    .isFunction ('fail', '.addImage', pastec && pastec.addImage)
    .isFunction ('fail', '.deleteImage', pastec && pastec.deleteImage)
    .done ();
});


/**
 * Opensource API
 */

dotest.add ('Opensource API - Method ping', function (test) {
  pastec.ping (function (err, data) {
    test (err)
      .isObject ('fail', 'data', data)
      .isExactly ('warn', 'data.type', data && data.type, 'PONG')
      .done ();
  });
});

dotest.add ('Opensource API - Method writeIndex', function (test) {
  pastec.writeIndex ('./sampleIndex.dat', function (err, data) {
    test (err)
      .isObject ('fail', 'data', data)
      .isExactly ('warn', 'data.type', data && data.type, 'INDEX_WRITTEN')
      .done ();
  });
});

dotest.add ('Opensource API - Method clearIndex', function (test) {
  pastec.clearIndex (function (err, data) {
    test (err)
      .isObject ('fail', 'data', data)
      .isExactly ('warn', 'data.type', data && data.type, 'INDEX_CLEARED')
      .done ();
  });
});

dotest.add ('Opensource API - Method loadIndex', function (test) {
  pastec.loadIndex ('./sampleIndex.dat', function (err, data) {
    test (err)
      .isObject ('fail', 'data', data)
      .isExactly ('warn', 'data.type', data && data.type, 'INDEX_LOADED')
      .done ();
  });
});

dotest.add ('Opensource API - Method addImage - filename', function (test) {
  pastec.addImage (path.join (dir, 'imageSample.jpg'), 999999999999999, function (err, data) {
    test (err)
      .isObject ('fail', 'data', data)
      .isExactly ('warn', 'data.type', data && data.type, 'IMAGE_ADDED')
      .done ();
  });
});

dotest.add ('Opensource API - Method addImage - buffer', function (test) {
  fs.readFile (path.join (dir, 'imageSample.jpg'), function (errFile, file) {
    test (errFile);

    pastec.addImage (file, 999999999999999, function (err, data) {
      test (err)
        .isObject ('fail', 'data', data)
        .isExactly ('warn', 'data.type', data && data.type, 'IMAGE_ADDED')
        .done ();
    });
  });
});

dotest.add ('Opensource API - Method addImage - error', function (test) {
  pastec.addImage ('', 999999999999999, function (err, data) {
    test ()
      .isError ('fail', 'err', err)
      .info (err.message)
      .isUndefined ('fail', 'data', data)
      .done ();
  });
});

dotest.add ('Opensource API - Method searchIndex - filename', function (test) {
  pastec.searchIndex (path.join (dir, 'imageSample.jpg'), function (err, data) {
    test (err)
      .isObject ('fail', 'data', data)
      .isExactly ('warn', 'data.type', data && data.type, 'SEARCH_RESULTS')
      .done ();
  });
});


dotest.add ('Opensource API - Method searchIndex - buffer', function (test) {
  fs.readFile (path.join (dir, 'imageSample.jpg'), function (errFile, file) {
    test (errFile);

    pastec.searchIndex (file, function (err, data) {
      test (err)
        .isObject ('fail', 'data', data)
        .isExactly ('warn', 'data.type', data && data.type, 'SEARCH_RESULTS')
        .done ();
    });
  });
});

dotest.add ('Opensource API - Method searchIndex - error', function (test) {
  pastec.searchIndex ('', function (err, data) {
    test ()
      .isError ('fail', 'err', err)
      .info (err.message)
      .isUndefined ('fail', 'data', data)
      .done ();
  });
});

dotest.add ('Opensource API - Method deleteImage', function (test) {
  pastec.deleteImage (999999999999999, function (err, data) {
    test (err)
      .isObject ('fail', 'data', data)
      .isExactly ('warn', 'data.type', data && data.type, 'IMAGE_REMOVED')
      .done ();
  });
});

dotest.add ('Opensource API - Error: invalid authkey', function (test) {
  const tmp = app ({
    endpoint: opensource.endpoint,
    indexid: opensource.indexid,
    authkey: '-',
    timeout: opensource.timeout
  });

  tmp.writeIndex ('./sampleIndex.dat', function (err, data) {
    test ()
      .isError ('fail', 'err', err)
      .isExactly ('fail', 'err.message', err && err.message, 'invalid authkey')
      .isExactly ('warn', 'err.error', err && err.error, 'AUTHENTIFICATION_ERROR')
      .isUndefined ('fail', 'data', data)
      .done ();
  });
});

dotest.add ('Opensource API - Error: request failed', function (test) {
  const tmp = app ({
    endpoint: opensource.endpoint,
    indexid: opensource.indexid,
    authkey: opensource.authkey,
    timeout: 1
  });

  tmp.writeIndex ('./sampleIndex.dat', function (err, data) {
    test ()
      .isError ('fail', 'err', err)
      .isExactly ('fail', 'err.message', err && err.message, 'request failed')
      .isError ('fail', 'err.error', err && err.error)
      .isExactly ('fail', 'err.error.code', err && err.error && err.error.code, 'TIMEOUT')
      .isUndefined ('fail', 'data', data)
      .done ();
  });
});


/**
 * Hosted API
 */

dotest.add ('Hosted API - Method ping', function (test) {
  pastec = app (hosted);

  pastec.ping (function (err, data) {
    test ()
      .isError ('fail', 'err', err)
      .isExactly ('fail', 'err.message', err && err.message, 'endpoint does not allow ping')
      .isUndefined ('fail', 'data', data)
      .done ();
  });
});

dotest.add ('Hosted API - Method writeIndex', function (test) {
  pastec.writeIndex ('./sampleIndex.dat', function (err, data) {
    test (err)
      .isObject ('fail', 'data', data)
      .isExactly ('warn', 'data.type', data && data.type, 'INDEX_WRITTEN')
      .done ();
  });
});

dotest.add ('Hosted API - Method clearIndex', function (test) {
  pastec.clearIndex (function (err, data) {
    test (err)
      .isObject ('fail', 'data', data)
      .isExactly ('warn', 'data.type', data && data.type, 'INDEX_CLEARED')
      .done ();
  });
});

dotest.add ('Hosted API - Method loadIndex', function (test) {
  pastec.loadIndex ('./sampleIndex.dat', function (err, data) {
    test (err)
      .isObject ('fail', 'data', data)
      .isExactly ('warn', 'data.type', data && data.type, 'INDEX_LOADED')
      .done ();
  });
});

dotest.add ('Hosted API - Method addImage - filename', function (test) {
  pastec.addImage (path.join (dir, 'imageSample.jpg'), 999999999999999, function (err, data) {
    test (err)
      .isObject ('fail', 'data', data)
      .isExactly ('warn', 'data.type', data && data.type, 'IMAGE_ADDED')
      .done ();
  });
});

dotest.add ('Hosted API - Method addImage - buffer', function (test) {
  fs.readFile (path.join (dir, 'imageSample.jpg'), function (errFile, file) {
    test (errFile);

    pastec.addImage (file, 999999999999999, function (err, data) {
      test (err)
        .isObject ('fail', 'data', data)
        .isExactly ('warn', 'data.type', data && data.type, 'IMAGE_ADDED')
        .done ();
    });
  });
});

dotest.add ('Hosted API - Method addImage - error', function (test) {
  pastec.addImage ('', 999999999999999, function (err, data) {
    test ()
      .isError ('fail', 'err', err)
      .info (err.message)
      .isUndefined ('fail', 'data', data)
      .done ();
  });
});

dotest.add ('Hosted API - Method searchIndex - filename', function (test) {
  pastec.searchIndex (path.join (dir, 'imageSample.jpg'), function (err, data) {
    test (err)
      .isObject ('fail', 'data', data)
      .isExactly ('warn', 'data.type', data && data.type, 'SEARCH_RESULTS')
      .done ();
  });
});


dotest.add ('Hosted API - Method searchIndex - buffer', function (test) {
  fs.readFile (path.join (dir, 'imageSample.jpg'), function (errFile, file) {
    test (errFile);

    pastec.searchIndex (file, function (err, data) {
      test (err)
        .isObject ('fail', 'data', data)
        .isExactly ('warn', 'data.type', data && data.type, 'SEARCH_RESULTS')
        .done ();
    });
  });
});

dotest.add ('Hosted API - Method searchIndex - error', function (test) {
  pastec.searchIndex ('', function (err, data) {
    test ()
      .isError ('fail', 'err', err)
      .info (err.message)
      .isUndefined ('fail', 'data', data)
      .done ();
  });
});

dotest.add ('Hosted API - Method deleteImage', function (test) {
  pastec.deleteImage (999999999999999, function (err, data) {
    test (err)
      .isObject ('fail', 'data', data)
      .isExactly ('warn', 'data.type', data && data.type, 'IMAGE_REMOVED')
      .done ();
  });
});

dotest.add ('Hosted API - Error: invalid authkey', function (test) {
  const tmp = app ({
    endpoint: hosted.endpoint,
    indexid: hosted.indexid,
    authkey: '-',
    timeout: hosted.timeout
  });

  tmp.writeIndex ('./sampleIndex.dat', function (err, data) {
    test ()
      .isError ('fail', 'err', err)
      .isExactly ('fail', 'err.message', err && err.message, 'invalid authkey')
      .isExactly ('warn', 'err.error', err && err.error, 'AUTHENTIFICATION_ERROR')
      .isUndefined ('fail', 'data', data)
      .done ();
  });
});

dotest.add ('Hosted API - Error: request failed', function (test) {
  const tmp = app ({
    endpoint: hosted.endpoint,
    indexid: hosted.indexid,
    authkey: hosted.authkey,
    timeout: 1
  });

  tmp.writeIndex ('./sampleIndex.dat', function (err, data) {
    test ()
      .isError ('fail', 'err', err)
      .isExactly ('fail', 'err.message', err && err.message, 'request failed')
      .isError ('fail', 'err.error', err && err.error)
      .isExactly ('fail', 'err.error.code', err && err.error && err.error.code, 'TIMEOUT')
      .isUndefined ('fail', 'data', data)
      .done ();
  });
});

dotest.add ('Setup arguments', function (test) {
  const tmp = app ('example.lan', 123);

  test ()
    .isExactly ('fail', 'config.endpoint', tmp.config.endpoint, 'example.lan')
    .isNull ('fail', 'config.authkey', tmp.config.authkey)
    .isExactly ('fail', 'config.timeout', tmp.config.timeout, 123)
    .done ();
});


// Start the tests
dotest.run ();


index.js
/*
Name:         pastecapi
Description:  Lightweight module for Pastec image recognition API
Author:       Franklin van de Meent (https://frankl.in)
Source code:  https://github.com/fvdm/nodejs-pastecapi
Feedback:     https://github.com/fvdm/nodejs-pastecapi/issues
License:      Unlicense (Public Domain, see LICENSE file)
*/

var httpreq = require ('httpreq');
var fs = require ('fs');

var config = {
  endpoint: 'http://localhost:4121',
  timeout: 5000
};


/**
 * Process response error
 *
 * @callback callback
 * @param message {string} - error.message
 * @param err {Error, string} - error.error
 * @param res {object} - httpreq response details
 * @param callback {function} - `function (err) {}`
 * @returns {void}
 */

function processError (message, err, res, callback) {
  var error = new Error (message);

  error.statusCode = res && res.statusCode;
  error.error = err;
  callback (error);
}


/**
 * Process HTTP response
 *
 * @callback callback
 * @param err {Error, null} - httpreq error
 * @param res {object} - httpreq response details
 * @param callback {function} - `function (err, data) {}`
 * @returns {void}
 */

function processResponse (err, res, callback) {
  var data = res && res.body || '';

  if (err) {
    processError ('request failed', err, res, callback);
    return;
  }

  try {
    data = JSON.parse (data);
  } catch (e) {
    processError ('request failed', e, res, callback);
    return;
  }

  if (data.type && data.type.match (/(AUTHENTIFICATION_ERROR|NO_AUTH_KEY|WRONG_INDEX_OR_AUTH_KEY)/)) {
    processError ('invalid authkey', data.type, res, callback);
    return;
  }

  callback (null, data);
}


/**
 * Communicate with API
 *
 * @callback callback
 * @param {object} options
 * @param {string} options.method=GET - HTTP method
 * @param {string} options.path=/ - Request path, i.e. /index/searcher
 * @param {object} options.parameters - Request query parameters
 * @param {object} options.json - POST JSON data
 * @param {object} options.files - Upload files paths
 * @param {number} options.timeout=5000 - Override request timeout in ms
 * @param {function} callback - Callback function
 * @returns {void}
 */

function talk (options, callback) {
  if (config.indexid) {
    options.path = options.path.replace (/^\/index/, '/indexes/' + config.indexid);

    if (options.json && options.json.type === 'PING' && config.endpoint === 'https://api.pastec.io') {
      callback (new Error ('endpoint does not allow ping'));
      return;
    }
  }

  options.url = config.endpoint + options.path;
  options.method = options.method || 'GET';
  options.timeout = options.timeout || config.timeout;
  options.headers = options.headers || {};
  options.headers ['User-Agent'] = 'pastecapi node/' + process.versions.node;

  if (config.authkey) {
    options.headers.AuthKey = config.authkey;
  }

  function doResponse (err, res) {
    processResponse (err, res, callback);
  }

  httpreq.doRequest (options, doResponse);
}


/**
 * Test API access
 *
 * @callback callback
 * @param {function} callback - Process result
 * @returns {void}
 */

function ping (callback) {
  var options = {
    method: 'POST',
    path: '/',
    json: {
      type: 'PING'
    }
  };

  talk (options, callback);
}


/**
 * Load index from file on server
 *
 * @param {string} indexPath - Path to .dat file on server
 * @param {function} callback - Process result
 * @returns {void}
 */

function loadIndex (indexPath, callback) {
  var options = {
    method: 'POST',
    path: '/index/io',
    json: {
      type: 'LOAD',
      index_path: indexPath
    }
  };

  talk (options, callback);
}


/**
 * Write index to file on server
 *
 * @callback callback
 * @param {string} indexPath - Path of .dat file on server
 * @param {function} callback - Process result
 * @returns {void}
 */

function writeIndex (indexPath, callback) {
  var options = {
    method: 'POST',
    path: '/index/io',
    json: {
      type: 'WRITE',
      index_path: indexPath
    }
  };

  talk (options, callback);
}


/**
 * Clear index on server
 *
 * @callback callback
 * @param {function} callback - Process result
 * @returns {void}
 */

function clearIndex (callback) {
  var options = {
    method: 'POST',
    path: '/index/io',
    json: {
      type: 'CLEAR'
    }
  };

  talk (options, callback);
}


/**
 * Match image against index
 *
 * @callback callback
 * @param {string|buffer} image - Buffer or local path to image
 * @param {function} callback - Process result
 * @returns {void}
 */

function searchIndex (image, callback) {
  var options = {
    method: 'POST',
    path: '/index/searcher',
    binary: true
  };

  // Binary pass through
  if (image instanceof Buffer) {
    options.body = image;
    options.headers = {
      'Content-Length': image.length,
      'Content-Type': 'image/jpeg'
    };

    talk (options, callback);
    return;
  }

  // Read file to buffer
  fs.readFile (image, function (err, data) {
    if (err) {
      callback (err);
      return;
    }

    options.body = data;
    options.headers = {
      'Content-Length': data.length,
      'Content-Type': 'image/jpeg'
    };

    talk (options, callback);
  });
}


/**
 * Remove image signature from index
 *
 * @callback callback
 * @param {number} imageId - Index ID of image signature
 * @param {function} callback - Process result
 * @returns {void}
 */

function deleteImage (imageId, callback) {
  var options = {
    method: 'DELETE',
    path: '/index/images/' + imageId
  };

  talk (options, callback);
}


/**
 * Add image signature to index
 *
 * @callback callback
 * @param {string|buffer} image - Buffer or local path to image
 * @param {number} imageId - Index ID for image
 * @param {function} callback - Process result
 * @returns {void}
 */

function addImage (image, imageId, callback) {
  var options = {
    method: 'PUT',
    path: '/index/images/' + imageId,
    binary: true
  };

  // Binary pass through
  if (image instanceof Buffer) {
    options.body = image;
    options.headers = {
      'Content-Length': image.length,
      'Content-Type': 'image/jpeg'
    };

    talk (options, callback);
    return;
  }

  // Read file to buffer
  fs.readFile (image, function (err, data) {
    if (err) {
      callback (err);
      return;
    }

    options.body = data;
    options.headers = {
      'Content-Length': data.length,
      'Content-Type': 'image/jpeg'
    };

    talk (options, callback);
  });
}


/**
 * Module setup
 *
 * @param {string} endpoint - REST API base URL
 * @param (string} authkey - API auth key (--auth-key flag)
 * @param {number} timeout - Request wait timeout in ms
 * @returns {object} - Module methods
 */

module.exports = function (endpoint, authkey, timeout) {
  if (typeof endpoint === 'object') {
    config.endpoint = endpoint.endpoint || config.endpoint;
    config.indexid = endpoint.indexid || null;
    config.authkey = endpoint.authkey || null;
    config.timeout = endpoint.timeout || config.timeout;
  } else {
    if (typeof authkey === 'number') {
      timeout = authkey;
      authkey = null;
    }

    config.endpoint = endpoint || config.endpoint;
    config.timeout = timeout || config.timeout;
    config.authkey = authkey || null;
  }

  return {
    ping,
    loadIndex,
    writeIndex,
    clearIndex,
    searchIndex,
    addImage,
    deleteImage
  };
};
